File: src/gullak/ledger/parser.py
"""Parser for ledger-cli format files."""

import contextlib
import re
from datetime import date
from decimal import Decimal, InvalidOperation
from pathlib import Path

from .models import Posting, Transaction, TransactionSource, TransactionStatus


class LedgerParser:
    """Parse ledger-format files into Transaction objects."""

    # Match transaction header: 2024/01/15 [*!] Payee name
    DATE_PATTERN = re.compile(r"^(\d{4}[-/]\d{2}[-/]\d{2})\s*([*!])?\s*(.+?)(?:\s*;\s*(.*))?$")

    # Match posting: account  amount [currency]
    POSTING_PATTERN = re.compile(r"^\s{2,}([A-Za-z][^\d]*?)\s{2,}([-\d,_.]+)\s*(\w+)?(?:\s*;.*)?$")

    # Match comment line
    COMMENT_PATTERN = re.compile(r"^\s*;\s*(.*)$")

    GULLAK_ID_PATTERN = re.compile(r"gullak:id\s+(\w+)")
    GULLAK_SOURCE_PATTERN = re.compile(r"gullak:source\s+(\w+)")
    GULLAK_USER_PATTERN = re.compile(r"gullak:user\s+(.+)$")

    TAG_PATTERN = re.compile(r"^(\w+):\s*(.+)$")

    POSTING_WITH_COMMENT_PATTERN = re.compile(
        r"^\s{2,}([A-Za-z][^\d]*?)\s{2,}([-\d,_.]+)\s*(\w+)?(?:\s*;\s*(.*))?$"
    )

    def parse_file(self, path: Path) -> list[Transaction]:
        """Parse a ledger file and return list of transactions."""
        if not path.exists():
            return []

        content = path.read_text()
        return self.parse_string(content)

    def parse_string(self, content: str) -> list[Transaction]:
        """Parse ledger content string into transactions."""
        transactions: list[Transaction] = []
        current_txn: dict | None = None
        current_comments: list[str] = []

        for line in content.split("\n"):
            # Skip empty lines
            if not line.strip():
                if current_txn is not None:
                    # End of transaction
                    txn = self._build_transaction(current_txn, current_comments)
                    if txn:
                        transactions.append(txn)
                    current_txn = None
                    current_comments = []
                continue

            # Check for transaction header
            if match := self.DATE_PATTERN.match(line):
                # Save previous transaction
                if current_txn is not None:
                    txn = self._build_transaction(current_txn, current_comments)
                    if txn:
                        transactions.append(txn)

                # Start new transaction
                current_txn = {
                    "date": self._parse_date(match.group(1)),
                    "status": self._parse_status(match.group(2)),
                    "payee": match.group(3).strip(),
                    "postings": [],
                }
                current_comments = []
                continue

            if current_txn is not None and (match := self.POSTING_WITH_COMMENT_PATTERN.match(line)):
                posting = self._parse_posting(match)
                if posting:
                    current_txn["postings"].append(posting)
                    inline_comment = match.group(4)
                    if inline_comment:
                        current_comments.append(inline_comment)
                continue

            if current_txn is not None and (match := self.POSTING_PATTERN.match(line)):
                posting = self._parse_posting(match)
                if posting:
                    current_txn["postings"].append(posting)
                continue

            # Check for comment within transaction
            if current_txn is not None and (match := self.COMMENT_PATTERN.match(line)):
                current_comments.append(match.group(1))

        # Handle last transaction
        if current_txn is not None:
            txn = self._build_transaction(current_txn, current_comments)
            if txn:
                transactions.append(txn)

        return transactions

    # Match account directives: account Expenses:Food:Groceries
    ACCOUNT_DIRECTIVE_PATTERN = re.compile(r"^account\s+(\S+)")

    def extract_accounts(self, path: Path) -> set[str]:
        """Extract unique account names from ledger file.

        Reads both account directives (account X) and accounts used in
        transaction postings.
        """
        accounts: set[str] = set()

        if not path.exists():
            return accounts

        # First pass: scan for account directives (fast line scan)
        content = path.read_text()
        for line in content.split("\n"):
            if match := self.ACCOUNT_DIRECTIVE_PATTERN.match(line):
                account = match.group(1).strip()
                accounts.add(account)
                parts = account.split(":")
                for i in range(1, len(parts)):
                    accounts.add(":".join(parts[:i]))

        # Second pass: accounts from transaction postings
        for txn in self.parse_string(content):
            for posting in txn.postings:
                accounts.add(posting.account)
                parts = posting.account.split(":")
                for i in range(1, len(parts)):
                    accounts.add(":".join(parts[:i]))

        return accounts

    def extract_payees(self, path: Path) -> set[str]:
        """Extract unique payee names from ledger file."""
        return {txn.payee for txn in self.parse_file(path)}

    def _parse_date(self, date_str: str) -> date:
        """Parse date string (YYYY/MM/DD or YYYY-MM-DD)."""
        normalized = date_str.replace("/", "-")
        return date.fromisoformat(normalized)

    def _parse_status(self, status_char: str | None) -> TransactionStatus:
        """Parse status character to enum."""
        if status_char == "*":
            return TransactionStatus.CLEARED
        elif status_char == "!":
            return TransactionStatus.PENDING
        return TransactionStatus.UNCLEARED

    def _parse_posting(self, match: re.Match) -> Posting | None:
        """Parse posting from regex match."""
        account = match.group(1).strip()
        amount_str = match.group(2).replace(",", "").replace("_", "")
        currency = match.group(3) or "INR"

        try:
            amount = Decimal(amount_str)
        except InvalidOperation:
            return None

        return Posting(account=account, amount=amount, currency=currency)

    def _build_transaction(self, data: dict, comments: list[str]) -> Transaction | None:
        if not data.get("postings"):
            return None

        gullak_id = None
        source = None
        source_user = None
        note = None
        tags: dict[str, str] = {}

        for comment in comments:
            if match := self.GULLAK_ID_PATTERN.search(comment):
                gullak_id = match.group(1)
                continue

            if match := self.GULLAK_SOURCE_PATTERN.search(comment):
                source_str = match.group(1).lower()
                with contextlib.suppress(ValueError):
                    source = TransactionSource(source_str)
                continue

            if match := self.GULLAK_USER_PATTERN.search(comment):
                source_user = match.group(1).strip()
                continue

            if match := self.TAG_PATTERN.match(comment):
                key, value = match.groups()
                if key != "gullak":
                    tags[key] = value
                continue

            if note is None:
                note = comment

        txn_data = {
            "date": data["date"],
            "payee": data["payee"],
            "status": data["status"],
            "postings": data["postings"],
            "note": note,
            "tags": tags,
        }

        if gullak_id:
            txn_data["gullak_id"] = gullak_id
        if source:
            txn_data["source"] = source
        if source_user:
            txn_data["source_user"] = source_user

        return Transaction(**txn_data)

File: pi-server/src/ledger/parser.ts
import { type LedgerTransaction, type Posting, type TransactionStatus } from "./models.js";

const DATE_PATTERN = /^(\d{4}[-/]\d{2}[-/]\d{2})\s*([*!])?\s*(.+?)(?:\s*;\s*(.*))?$/;
const COMMENT_PATTERN = /^\s*;\s*(.*)$/;
const POSTING_LINE_PATTERN = /^\s{2,}(.+?)(?:\s{2,}(.+?))?(?:\s*;\s*(.*))?$/;
const ACCOUNT_DIRECTIVE_PATTERN = /^account\s+(\S+)/;
const GULLAK_ID_PATTERN = /gullak:id\s+(\w+)/;
const GULLAK_SOURCE_PATTERN = /gullak:source\s+(\w+)/;
const GULLAK_USER_PATTERN = /gullak:user\s+(.+)$/;
const TAG_PATTERN = /^(\w+):\s*(.+)$/;

interface WorkingTransaction {
  date: string;
  payee: string;
  status: TransactionStatus;
  postings: Posting[];
  missingPostings: Array<{ account: string; currency: string }>;
}

function parseAmount(raw: string): number | undefined {
  const normalized = raw.replaceAll(",", "").replaceAll("_", "");
  const value = Number.parseFloat(normalized);
  return Number.isFinite(value) ? Number(value.toFixed(2)) : undefined;
}

function parseStatus(raw: string | undefined): TransactionStatus {
  if (raw === "*" || raw === "!") {
    return raw;
  }

  return "";
}

function buildTransaction(
  current: WorkingTransaction | undefined,
  comments: string[],
): LedgerTransaction | undefined {
  if (!current || (current.postings.length === 0 && current.missingPostings.length === 0)) {
    return undefined;
  }

  if (current.missingPostings.length === 1 && current.postings.length > 0) {
    const missing = current.missingPostings[0];
    const balance = Number(
      (-current.postings.reduce((sum, posting) => sum + posting.amount, 0)).toFixed(2),
    );
    current.postings.push({
      account: missing.account,
      amount: balance,
      currency: missing.currency || current.postings[0]?.currency || "INR",
    });
  } else if (current.missingPostings.length > 0) {
    for (const missing of current.missingPostings) {
      current.postings.push({
        account: missing.account,
        amount: 0,
        currency: missing.currency || current.postings[0]?.currency || "INR",
      });
    }
  }

  let gullakId = "";
  let source: LedgerTransaction["source"];
  let sourceUser: string | undefined;
  let note: string | undefined;
  const tags: Record<string, string> = {};

  for (const comment of comments) {
    const idMatch = comment.match(GULLAK_ID_PATTERN);
    if (idMatch) {
      gullakId = idMatch[1];
      continue;
    }

    const sourceMatch = comment.match(GULLAK_SOURCE_PATTERN);
    if (sourceMatch) {
      source = sourceMatch[1] as LedgerTransaction["source"];
      continue;
    }

    const userMatch = comment.match(GULLAK_USER_PATTERN);
    if (userMatch) {
      sourceUser = userMatch[1].trim();
      continue;
    }

    const tagMatch = comment.match(TAG_PATTERN);
    if (tagMatch && tagMatch[1] !== "gullak") {
      tags[tagMatch[1]] = tagMatch[2];
      continue;
    }

    if (!note) {
      note = comment;
    }
  }

  return {
    date: current.date.replaceAll("/", "-"),
    payee: current.payee,
    status: current.status,
    postings: current.postings,
    note,
    tags,
    gullakId,
    source,
    sourceUser,
  };
}

export function parseLedger(content: string): LedgerTransaction[] {
  const transactions: LedgerTransaction[] = [];
  let current: WorkingTransaction | undefined;
  let comments: string[] = [];

  for (const line of content.split("\n")) {
    if (!line.trim()) {
      const built = buildTransaction(current, comments);
      if (built) {
        transactions.push(built);
      }
      current = undefined;
      comments = [];
      continue;
    }

    const headerMatch = line.match(DATE_PATTERN);
    if (headerMatch) {
      const previous = buildTransaction(current, comments);
      if (previous) {
        transactions.push(previous);
      }

      current = {
        date: headerMatch[1].replaceAll("/", "-"),
        status: parseStatus(headerMatch[2]),
        payee: headerMatch[3].trim(),
        postings: [],
        missingPostings: [],
      };
      comments = [];
      continue;
    }

    if (!current) {
      continue;
    }

    const commentMatch = line.match(COMMENT_PATTERN);
    if (commentMatch) {
      comments.push(commentMatch[1]);
      continue;
    }

    const postingMatch = line.match(POSTING_LINE_PATTERN);
    if (postingMatch) {
      const account = postingMatch[1].trim();
      const posting = parsePosting(postingMatch[2]);
      if (posting.amount !== undefined) {
        current.postings.push({
          account,
          amount: posting.amount,
          currency: posting.currency,
        });
      } else {
        current.missingPostings.push({ account, currency: posting.currency });
      }

      if (postingMatch[3]) {
        comments.push(postingMatch[3]);
      }
    }
  }

  const last = buildTransaction(current, comments);
  if (last) {
    transactions.push(last);
  }

  return transactions;
}

function parsePosting(rawAmount: string | undefined): { amount?: number; currency: string } {
  if (!rawAmount) {
    return { currency: "INR" };
  }

  const tokens = rawAmount.trim().split(/\s+/).filter(Boolean);
  const amountToken = tokens.find((token) => /^[-\d,_.]+$/.test(token));
  const amount = amountToken ? parseAmount(amountToken) : undefined;
  const amountIndex = amountToken ? tokens.indexOf(amountToken) : -1;
  const currencyCandidate = amountIndex >= 0 ? tokens[amountIndex + 1] : undefined;

  return {
    amount,
    currency:
      currencyCandidate && !currencyCandidate.startsWith("@") && !currencyCandidate.startsWith("{")
        ? currencyCandidate
        : "INR",
  };
}

export function extractAccounts(content: string): string[] {
  const accounts = new Set<string>();

  for (const line of content.split("\n")) {
    const match = line.match(ACCOUNT_DIRECTIVE_PATTERN);
    if (!match) {
      continue;
    }

    const parts = match[1].split(":");
    for (let index = 1; index <= parts.length; index += 1) {
      accounts.add(parts.slice(0, index).join(":"));
    }
  }

  for (const transaction of parseLedger(content)) {
    for (const posting of transaction.postings) {
      const parts = posting.account.split(":");
      for (let index = 1; index <= parts.length; index += 1) {
        accounts.add(parts.slice(0, index).join(":"));
      }
    }
  }

  return [...accounts].sort();
}

export function findTransactionSpan(
  lines: string[],
  gullakId: string,
): { start: number; end: number } | undefined {
  const marker = `gullak:id ${gullakId}`;
  let markerLine = -1;

  for (let index = 0; index < lines.length; index += 1) {
    if (lines[index].includes(marker)) {
      markerLine = index;
      break;
    }
  }

  if (markerLine === -1) {
    return undefined;
  }

  let start = markerLine;
  for (let index = markerLine - 1; index >= 0; index -= 1) {
    if (!lines[index].trim()) {
      break;
    }

    if (DATE_PATTERN.test(lines[index])) {
      start = index;
      break;
    }
  }

  let end = markerLine + 1;
  for (let index = markerLine + 1; index < lines.length; index += 1) {
    if (!lines[index].trim()) {
      end = index;
      break;
    }
    end = index + 1;
  }

  return { start, end };
}
